### ISSUE TITLE
Possible unsound code in wasm-bindgen-futures

### ISSUE THREAD

**Issue by awestlake87** [rustwasm/wasm-bindgen#1881]
### Describe the Bug

It looks like wasm-bindgen-futures might have some unsound code. I've been noticing some weirdness when writing a custom poll function. I've worked with async/await before, but this is the first time I've tried implementing a future with the trait, so maybe I'm doing something horribly wrong.

### Steps to Reproduce
This is a pretty long piece of code (sorry), but it demonstrates the problem and isolates it to a specific test case.

```rust
use std::{
    future::Future,
    mem,
    pin::Pin,
    task::{Context, Poll},
};

use futures::{channel::mpsc, sink::SinkExt, stream::StreamExt};

use wasm_bindgen_test::wasm_bindgen_test;

// async-std does not compile for WASM yet.
#[cfg(not(target_arch = "wasm32"))]
use async_std::task;

// A simple task distilled down to something that sends and receives an item on
// an mpsc channel. Pretty dumb, I know, but it fails for some reason under
// certain circumstances.
async fn test_task(num: i32) -> i32 {
    let (mut tx, mut rx) = mpsc::channel(1);

    tx.send(num).await.unwrap();

    rx.next().await.unwrap()
}

// Wraps a list of tasks and polls to collect their results. This is very
// similar to the 'join' combinator, but in my own project, I'm trying to
// do something like a dynamic join where I add tasks to the list between
// polls. I just isolated the bits that cause the problem here, though.
struct TaskWrapper {
    tasks: Vec<Pin<Box<dyn Future<Output = i32>>>>,
    results: Vec<Option<i32>>,
}

impl TaskWrapper {
    fn new(tasks: Vec<Pin<Box<dyn Future<Output = i32>>>>) -> Self {
        let mut results = Vec::with_capacity(tasks.len());
        results.resize(tasks.len(), None);

        Self { tasks, results }
    }
}

impl Future for TaskWrapper {
    type Output = Vec<i32>;

    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
        let mut still_polling = false;

        let mut results = mem::replace(&mut self.results, vec![]);

        // poll each task and store the result if it's ready
        for (i, task) in self.tasks.iter_mut().enumerate() {
            // check if we should poll this task
            if results[i].is_none() {
                still_polling = true;

                match task.as_mut().poll(cx) {
                    Poll::Pending => continue,
                    Poll::Ready(result) => results[i] = Some(result),
                }
            }
        }

        mem::replace(&mut self.results, results);

        if still_polling {
            Poll::Pending
        } else {
            // unwrap each result and collect into a vec
            Poll::Ready(self.results.drain(..).map(|res| res.unwrap()).collect())
        }
    }
}

// This is the problematic task in WASM
async fn unsound_in_wasm_task() {
    let wrapper = TaskWrapper::new(vec![
        Pin::from(Box::new(test_task(12))),
        Pin::from(Box::new(test_task(43))),
        Pin::from(Box::new(test_task(-15))),
    ]);
    assert_eq!(wrapper.await, vec![12, 43, -15]);
}

// Using the executor in async-std, the task runs just fine.
#[cfg(not(target_arch = "wasm32"))]
#[test]
fn sound_in_async_std() {
    task::block_on(unsound_in_wasm_task());
}

// The same task on the wasm-bindgen-futures executor crashes
#[wasm_bindgen_test(async)]
async fn unsound_in_wasm() {
    unsound_in_wasm_task().await;
}

// Just a small proof that the test_task works on its own
#[wasm_bindgen_test(async)]
async fn this_works_in_wasm() {
    assert_eq!(test_task(12).await, 12);
}
```

You can clone the full repo [here](https://github.com/awestlake87/wasm-bug).

I run this code successfully with async-std using the following command:
```bash
$ cargo test
```

When I run the code with wasm-pack, it fails:
```bash
$ wasm-pack test --node
```

### Expected Behavior

So, I expect all of these tests to pass. The one that I compile for async std works perfectly fine, but the wasm-pack test fails in a very concerning way.

### Actual Behavior

Here's the output of `cargo test`. The test compiles and passes:
```bash
wasm-bug$ cargo test
   Compiling wasm-bug v0.1.0 (Documents/projects/wasm-bug)
    Finished test [unoptimized + debuginfo] target(s) in 1.24s
     Running target/debug/deps/wasm_bug-5fb538b0a26c3d50

running 1 test
test tests::sound_in_async_std ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests wasm-bug

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

The wasm-pack test fails in a subtle, but very concerning way:
```bash
wasm-bug$ wasm-pack test --node
[INFO]: Checking for the Wasm target...
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
[INFO]: Installing wasm-bindgen...
    Finished test [unoptimized + debuginfo] target(s) in 0.03s
     Running target/wasm32-unknown-unknown/debug/deps/wasm_bug-21a10e21fafbd517.wasm
running 2 tests                                   

test wasm_bug::tests::this_works_in_wasm ... ok
wasm-bug$ 
```
This is interesting because it says `running 2 tests`, but only runs 1 and completely quits. Now I only have 2 tests here, but in my other project I have several. The wasm-bindgen test runner has failed out completely on this test and will not run any more after that. I'm not sure if this is a panic somewhere or if it's some sort of memory violation, but no error message, exit code 0. It's all very strange, and I'm not sure what to make of it.

### Additional Context

WASM is pretty bleeding edge, so as far as I'm aware, there are not really any good options for debugging past this point, so I'm kinda stuck. If there are, please let me know. I've seen some other odd quirks but have not been able to isolate them just yet and I would gladly help to get them fixed if I can. I love the direction Rust + async/await + WASM has been going. You guys are doing some awesome work!

Here's some versioning info for you guys in case you have some trouble reproducing it:
```
wasm-bug$ wasm-pack --version
wasm-pack 0.8.1
```
 
```
wasm-bug$ node --version
v12.13.0
```

```
wasm-bug$ cargo --version
cargo 1.41.0-nightly (8280633db 2019-11-11)
```

```
wasm-bug$ rustc --version
rustc 1.41.0-nightly (412f43ac5 2019-11-24)
```

**Comment by Pauan** [comment]
Just a note that you can use `FuturesOrdered` or `FuturesUnordered` which is similar to your `TaskWrapper`.

Also you can use `vec![None; tasks.len()]` in `TaskWrapper::new`. That's a lot nicer than using `Vec::with_capacity` and `resize`.

And you can use `zip` inside of the `poll` implementation, rather than all the shenanigans with `mem::replace` and `enumerate` and such. That will also be faster.

As for debugging, you usually would use [`console_error_panic_hook`](https://crates.io/crates/console_error_panic_hook), but that's usually used in `main`, not in a test. And you can use [`web-sys`](https://crates.io/crates/web-sys) for logging to the console, which is also useful for debugging. The [`log`](https://crates.io/crates/log) crate also [works with wasm](https://crates.io/crates/console_log).

As for the bug itself, it seems to be a bug in your code. What's happening is:

1. On the first tick the `results` is `[None, None, None]`

2. Then it polls the Futures, and all of the Futures are done, so now `results` is `[Some(12), Some(43), Some(-15)]`

3. But `still_polling` is `true`, because your check is wrong, so it returns `Poll::Pending`.

4. But all the Futures are done, so they never wake up again, so `TaskWrapper` will just remain as `Poll::Pending` forever (in other words, an infinite hang).

An infinitely hanging Future isn't unsound, it is perfectly valid (there's even [a function that does that](https://docs.rs/futures/0.3.1/futures/future/fn.pending.html)), which is why async code needs to have timeouts.

But obviously you don't want your Future to hang. The hanging isn't caused by wasm-bindgen, instead your `still_polling` logic is wrong. You need to change it to this:

```rust
// poll each task and store the result if it's ready
for (i, task) in self.tasks.iter_mut().enumerate() {
    // check if we should poll this task
    if results[i].is_none() {
        match task.as_mut().poll(cx) {
            Poll::Pending => {
                still_polling = true;
                continue;
            },
            Poll::Ready(result) => {
                results[i] = Some(result);
            },
        }
    }
}
```

Note that it only sets `still_polling` if the Future is `Pending`.

As for why your `TaskWrapper` works in `async-std`, it seems that `async-std` keeps polling `TaskWrapper` even though it returned `Poll::Pending`.

It's valid for `async-std` to do that, but you can't rely on it (as you found out, wasm-bindgen only re-polls the Future when it wakes up, which is *also* valid, and more efficient).

**Comment by awestlake87** [comment]
Ok, I tried your suggestions out and it worked! Thanks for the help and the amazing response time!

You were right about the `still_polling` check. I guess that's because I didn't have a good understanding of what Poll::Pending means. I thought it meant 'I'm still waiting on a result, poll me later to see if it's ready'. If I understand you right, it means 'I'm waiting on some event to wake me up, don't bother me again until it's ready'.

So now I think I understand why the future was hanging, however I'm still a bit fuzzy on what happened with the test runner. It looks like node just exited when this future was hung, which was what made me think there was some sort of crash. Did node exit just because it didn't think it had any work left on the event loop?

**Comment by awestlake87** [comment]
Oh, and [FuturesUnordered](https://docs.rs/futures/0.3.1/futures/stream/struct.FuturesUnordered.html) sounds like exactly what I wanted. I could have saved myself a lot of time if I had just dug further into the futures-rs docs.

Again, thanks for your help @Pauan! I've been spending too much time in C++ world. Any time something exits early I just assume it crashed horribly :P.

Looking forward to your work on [gloo](https://github.com/rustwasm/gloo) by the way. I want that modular vdom!

**Comment by Pauan** [comment]
> I guess that's because I didn't have a good understanding of what Poll::Pending means. I thought it meant 'I'm still waiting on a result, poll me later to see if it's ready'. If I understand you right, it means 'I'm waiting on some event to wake me up, don't bother me again until it's ready'.

Yes, that's correct. When your Future returns `Poll::Pending`, the executor will wait for you to call `Waker::wake` (or `wake_by_ref`). Until you wake it up, it won't do anything.

An executor is *allowed* to spuriously poll the Future even if the Future hasn't called `Waker::wake` (that's what `async-std` is doing), but you can't rely on that.

If you want the behavior of "periodically poll me to see if I'm ready" you will have to arrange that yourself, by having the Future call `Waker::wake` at some point in the future, for example using a timer.

> Did node exit just because it didn't think it had any work left on the event loop?

Yes, [the behavior of Node](https://nodejs.org/api/process.html#process_event_exit) is that if all of the foreground JS has finished running, and there aren't any background JS tasks queued on the event loop (like `setTimeout` or Promises or things like that), then it will exit. Your code doesn't schedule any JS tasks (it's all pure Rust), so Node will immediately exit.

> I've been spending too much time in C++ world. Any time something exits early I just assume it crashed horribly :P.

Yeah, it's quite rare for a crash in Rust to *not* have a panic message associated with it. It's definitely one of the nice benefits of Rust.
